### PR TITLE
[M1132] Ingest templates and PIT & Fish Belt serializers

### PR DESCRIPTION
**Pull request by gridcell**

### CODE REVIEW

**Comment by gridcell** [comment]
[draft observation schema](https://trello.com/c/ECwCFYxn/1132-draft-observation-schema)

File: src/api/ingest/serializers.py [hunk 1]
```diff
@@ -0,0 +1,511 @@
...
+    _reverse_choices = {}
+
+    # PROJECT RELATED CHOICES
+    project_choices = {
+        "data__sample_event__site": None,
+        "data__sample_event__management": None,
+    }
+
+    # FIELDS
+    id = serializers.UUIDField(format="hex_verbose")
```
Comment (saanobhaai) [inline src/api/ingest/serializers.py]: Haven't seen us need to use `hex_verbose` before... Is this a not the default any more?

File: src/api/ingest/serializers.py [hunk 2]
```diff
@@ -0,0 +1,511 @@
...
+        child=serializers.CharField(), allow_empty=False
+    )
+
+    class Meta:
+        list_serializer_class = CollectRecordCSVListSerializer
+
+    def __init__(self, data=None, instance=None, project_choices=None, **kwargs):
+        if instance is not None:
+            raise NotImplementedError("instance argument not implemented")
+        if isinstance(data, Mapping):
```
Comment (saanobhaai) [inline src/api/ingest/serializers.py]: Where would `Mapping` be used?

File: src/api/ingest/serializers.py [hunk 3]
```diff
@@ -0,0 +1,511 @@
...
+    def formatted_errors(self):
+        errors = self.errors
+        return self.format_error(errors)
+
+
+class BenthicPITCSVSerializer(CollectRecordCSVSerializer):
+    protocol = "benthicpit"
+    observations_field = "data__obs_benthic_pits"
+    ordering_field = "data__obs_benthic_pits__interval"
+    header_map = CollectRecordCSVSerializer.header_map
```
Comment (saanobhaai) [inline src/api/ingest/serializers.py]: Can't this be `self.header_map`?

File: src/api/ingest/serializers.py [hunk 4]
```diff
@@ -0,0 +1,511 @@
...
+    )
+    data__sample_event__notes = serializers.CharField(required=False, allow_blank=True)
+
+    # data__observers = ProjectProfileSerializer(many=True)
+    data__observers = serializers.ListField(
+        child=serializers.CharField(), allow_empty=False
+    )
+
+    class Meta:
+        list_serializer_class = CollectRecordCSVListSerializer
```
Comment (saanobhaai) [inline src/api/ingest/serializers.py]: interesting... so how is `many_init` forced to be always true? If I called a subclass of this without `many=True` what would happen?

File: src/api/ingest/serializers.py [hunk 5]
```diff
@@ -0,0 +1,511 @@
...
+        required=False,
+        allow_null=True,
+        allow_blank=True,
+    )
+    data__sample_event__tide = serializers.ChoiceField(
+        choices=tide_choices, required=False, allow_null=True, allow_blank=True
+    )
+    data__sample_event__notes = serializers.CharField(required=False, allow_blank=True)
+
+    # data__observers = ProjectProfileSerializer(many=True)
```
Comment (saanobhaai) [inline src/api/ingest/serializers.py]: interesting -- but if the `child` approach works, delete comment

File: src/api/ingest/serializers.py [hunk 6]
```diff
@@ -0,0 +1,511 @@
...
+            [
+                (field_name, field.get_value(self._original_data))
+                for field_name, field in self.fields.items()
+                if (field.get_value(self._original_data) == empty)
+                and not field.read_only
+            ]
+        )
+
+    def validate(self, data):
+        # Validate common Transect level fields
```
Comment (saanobhaai) [inline src/api/ingest/serializers.py]: Why doesn't this call `super()`?

File: src/api/ingest/serializers.py (hunk 1 above)
Comment (gridcell) [reply src/api/ingest/serializers.py]: just being explicit, there are 4 options it can be

File: src/api/ingest/serializers.py (hunk 2 above)
Comment (gridcell) [reply src/api/ingest/serializers.py]: If you only passed in 1 record

File: src/api/ingest/serializers.py (hunk 3 above)
Comment (gridcell) [reply src/api/ingest/serializers.py]: self isn't defined yet, header_map is first inheriting `header_map` from `CollectRecordCSVSerializer` and then extending on the next line

File: src/api/ingest/README.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+- `project_choices`: pre-fetched choices that require a project id, example: ProjectProfile choices.
+- `error_row_offset`: An offset to add to $row_number value in formatted validation errors.  Default: 1
+
+
+### Other
+
+- Records created by serializer are batched, NO signals will be triggered.
+
+
+## Example Using Ingest Serializer
```
[the remark this answers is not in the document]
Comment (saanobhaai) [reply src/api/ingest/README.md]: Never minda -- separate ticket

File: src/api/ingest/serializers.py (hunk 4 above)
Comment (gridcell) [reply src/api/ingest/serializers.py]: Added to README that it must be True

**Pull request merged**